azkaban-common/src/main/bash/internal/startup-shared.sh

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env bash
+
+#---
+# parse_args:              parses all given args and handles args that control the behavior of
+#                          launching. All other unrecognized args will be sent to the process
+#                          to be launched at the end.
+#                          Note this function is not supposed to be run in a sub shell, otherwise
+#                          the vars won't be correctly exported to the child process, which is why
+#                          it does not attempt to return the value via stdout.
+# args:                    the name of the var to set final command to, the original command
+#---
+function parse_args {
+  local target_arg="$1"
+  shift
+  local final_command=''
+  for arg in "$@"
+  do
+    case "$arg" in
+      -f | --foreground)
+      # azkaban will run as a foreground process
+      export RUN_IN_FOREGROUND='true'
+      ;;
+      --no-fork)
```

Comment:
@xunnanxu `--no-fork` should be required if you are planning to run this in a containerized environment. 
The reason is sigterms from docker or other containers would go to the parent process, and it would be entirely up to the parent process to decide whether to propagate the signal to the children. 
For example, bash does not pass signal to children. If you send a sigterm to bash, it will simply exit, and send a sigkill to the children. This prevents the actual forked process(JVM or others) to get gracefully shutdown.

Coming up with a nice parent process that handles all the different signals and orphaned subprocesses is difficult. OS init(pid 1) does that, but it's not advisable to run init in a container. There are alternative implementations, such as dumb_init from Yelp and my_init from baseimage-docker, but both have some cons, and as you say in your comments, this would put additional burden on the users.
`--no-fork` on the other hand gives the user a very simple model. Container starts the main java process as the main process, and sends it all the signals, and container dies when the java process dies.
IMHO we should start with that simple model and let the user community decide if a more complex model is necessary.

Reply:
Hi @ashahab, hmm I'm not quite sure if you are proposing something new here or just repeating my suggestion in this PR. In fact I think "no fork" should be the default option (see new discussion in  #1669). HOWEVER, the current default behavior is to run it in background. I don't want to change that because users might be relying on that behavior already so it's not something we should be changing in 3.x. This PR just gives other users (like me) an option to run stuff differently.